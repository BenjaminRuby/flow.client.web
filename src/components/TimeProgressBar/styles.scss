@import "~@boomerang/boomerang-styles/lib/colors";
@import "~Styles/variables";

.c-time-progress-bar {
  height: 6.25rem;
  width: 100%;
  padding: 2rem 1.25rem 1.25rem 1.25rem;
  background-color: $white;
  border-radius: 0.25rem;
  grid-area: progress;
  display: flex;
  align-items: center;

  //override default styles
  .bmrg--b-tooltip {
    max-width: initial;
    opacity: 1;
  }
}

.b-time-progress-bar {
  position: relative;
  height: 2rem;
  width: 100%;
  border-radius: 0.25rem;
  background-color: $gray-20;
}

.b-time-progress-bar__start {
  position: absolute;
  font-size: 1rem;
  font-weight: 600;
  top: -2rem;
  left: 0;
}

.b-time-progress-bar__finish {
  position: absolute;
  font-size: 1rem;
  font-weight: 600;
  top: -2rem;
  right: 0;
}

.b-time-progress-bar__fillers {
  position: relative;
  height: 100%;
  display: flex;
  background: white;
}
