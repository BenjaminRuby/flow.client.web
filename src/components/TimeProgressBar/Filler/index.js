import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import moment from "moment";
import Tooltip from "@boomerang/boomerang-components/lib/Tooltip";
import { ACTIVITY_STATUSES, ACTIVITY_STATUSES_TO_TEXT } from "Constants/activityStatuses";
import getHumanizedDuration from "@boomerang/boomerang-utilities/lib/getHumanizedDuration";
import "./styles.scss";

Filler.propTypes = {
  backgroundColor: PropTypes.string.isRequired,
  duration: PropTypes.number.isRequired,
  finishTime: PropTypes.number.isRequired,
  id: PropTypes.string.isRequired,
  isFirstTask: PropTypes.bool,
  isLastTask: PropTypes.bool,
  percentOfTotal: PropTypes.number.isRequired,
  status: PropTypes.string.isRequired,
  taskId: PropTypes.string.isRequired,
  taskName: PropTypes.string.isRequired,
  updateActiveNode: PropTypes.func.isRequired
};

function Filler({
  duration,
  finishTime,
  id,
  isFirstTask, //TODO: for styling
  isLastTask, //TODO
  percentOfTotal,
  status,
  taskId,
  taskName,
  updateActiveNode
}) {
  const styles = {
    backgroundColor: status === ACTIVITY_STATUSES.COMPLETED ? "#57d785" : "#e06768",
    width: `${percentOfTotal}%`
  };

  const finishTimeFormatted = moment.unix(finishTime / 1000).format("hh:mm:ss a", { trim: false });
  return (
    <div className="c-time-progress-bar-filler" style={styles} data-tip data-for={id}>
      <button
        onClick={() => updateActiveNode(taskId)}
        className={classnames("b-time-progress-bar-filler", { "--first": isFirstTask, "--last": isLastTask })}
      />
      {percentOfTotal > 5 && <div className="b-time-progress-bar-filler__time">{finishTimeFormatted}</div>}

      <Tooltip id={id} place="bottom">
        <div className="c-time-progress-bar-filler-info">
          <section className="b-time-progress-bar-filler-info">
            <p className="b-time-progress-bar-filler-info__label">Task name:</p>
            <p className="b-time-progress-bar-filler-info__label">Duration:</p>
            <p className="b-time-progress-bar-filler-info__label">Status:</p>
          </section>
          <section className="b-time-progress-bar-filler-info">
            <p className="b-time-progress-bar-filler-info__value">{taskName}</p>
            <p className="b-time-progress-bar-filler-info__value">
              {getHumanizedDuration(Math.round(parseInt(duration / 1000), 10))}
            </p>
            <p className="b-time-progress-bar-filler-info__value">{ACTIVITY_STATUSES_TO_TEXT[status]}</p>
          </section>
        </div>
      </Tooltip>
    </div>
  );
}

export default Filler;
