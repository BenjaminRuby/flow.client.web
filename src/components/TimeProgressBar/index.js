import React from "react";
import PropTypes from "prop-types";
import orderBy from "lodash/orderBy";
import Filler from "./Filler";
import { EXECUTION_STATUSES } from "Constants/workflowExecutionStatuses";
import "./styles.scss";

const TimeProgressBar = ({ workflowExecution, updateActiveNode }) => {
  const steps = orderBy(workflowExecution.steps, ["order"], ["asc"]).filter(
    step =>
      step.flowTaskStatus !== EXECUTION_STATUSES.NOT_STARTED && step.flowTaskStatus !== EXECUTION_STATUSES.IN_PROGRESS
  );
  let durationSum = 0;
  steps.forEach(step => (durationSum += step.duration));

  return (
    <div className="c-time-progress-bar">
      <div className="b-time-progress-bar">
        <div className="b-time-progress-bar__start">Start</div>
        <div className="b-time-progress-bar__finish">Finish</div>
        <div className="b-time-progress-bar__fillers">
          {steps.map((step, index) => (
            <Filler
              duration={step.duration}
              finishTime={step.startTime + step.duration}
              id={step.id}
              key={step.id}
              percentOfTotal={(step.duration / durationSum) * 100}
              status={step.flowTaskStatus}
              taskId={step.taskId}
              taskName={step.taskName}
              updateActiveNode={updateActiveNode}
            />
          ))}
        </div>
      </div>
    </div>
  );
};

TimeProgressBar.propTypes = {
  workflowExecution: PropTypes.object.isRequired,
  updateActiveNode: PropTypes.func.isRequired
};

export default TimeProgressBar;
