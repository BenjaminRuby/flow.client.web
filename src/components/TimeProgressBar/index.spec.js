import React from "react";
import TimeProgressBar from "./index";

const mockfn = jest.fn();

const workflowExecution = {
  creationDate: 1540912389131,
  duration: 7178,
  id: "5bd8750446e4bf790fe60aef",
  status: "completed",
  workflowId: "5bd6fdf65a5df954ad47bdbe",
  workflowRevisionid: "5bd85a2146e4bf444d97417d",
  steps: [
    {
      activityId: "5bd8750446e4bf790fe60aef",
      duration: 2178,
      flowTaskStatus: "inProgress",
      id: "5bd8750546e4bf790fe60af0",
      order: 0,
      startTime: 1540912389131,
      taskId: "5006c70f-c7f0-4679-940c-82ba89755599",
      taskName: "task1"
    },
    {
      activityId: "5bd8750446e4bf790fe60aef",
      duration: 3000,
      flowTaskStatus: "notstarted",
      id: "5bd8750546e4bf790fe60af1",
      order: 1,
      startTime: 1540912391309,
      taskId: "d64fb25d-932a-4894-9087-ded1b63eb672",
      taskName: "task2"
    }
  ]
};

describe("TimeProgressBar --- Snapshot", () => {
  it("Capturing Snapshot of TimeProgressBar", () => {
    const { baseElement } = rtlRender(
      <TimeProgressBar workflowExecution={workflowExecution} updateActiveNode={mockfn} />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
