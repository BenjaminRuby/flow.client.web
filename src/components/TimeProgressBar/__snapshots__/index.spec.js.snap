// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeProgressBar --- Snapshot Capturing Snapshot of TimeProgressBar 1`] = `
<div
  className="c-time-progress-bar"
>
  <div
    className="b-time-progress-bar"
  >
    <div
      className="b-time-progress-bar__start"
    >
      Start
    </div>
    <div
      className="b-time-progress-bar__finish"
    >
      Finish
    </div>
  </div>
</div>
`;
